// @nolebase/vitepress-plugin-thumbnail-hash/dist/vite/index.mjs
import { join, relative } from 'node:path';
import { mkdir, rm, readFile, writeFile } from 'node:fs/promises';
import { normalizePath } from 'vite';
import { rgbaToThumbHash, thumbHashToDataURL } from 'thumbhash';
import { loadImage, createCanvas } from '@napi-rs/canvas';
import { cyan, gray } from 'colorette';
import ora from 'ora';
import { glob } from 'glob';
import { subtle } from 'uncrypto';

function binaryToBase64(binary) {
  return btoa(String.fromCharCode(...binary));
}
async function digestUint8ArrayDataSha256(data) {
  const hashBuffer = await subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(hashBuffer));
}
async function hash(data, length = 10) {
  const hashResult = await digestUint8ArrayDataSha256(data);
  const hashBase64 = binaryToBase64(new Uint8Array(hashResult));
  if (length > 0)
    return hashBase64.substring(0, length);
  return hashBase64;
}
function normalizeBase64(base64) {
  return base64.replace("/", "_").replace("+", "-").replace("=", "-");
}

async function calculateThumbHashForFile(imageData) {
  const image = await loadImage(imageData);
  const width = image.width;
  const height = image.height;
  const scale = 100 / Math.max(width, height);
  const resizedWidth = Math.round(width * scale);
  const resizedHeight = Math.round(height * scale);
  const canvas = createCanvas(resizedWidth, resizedHeight);
  const context = canvas.getContext("2d");
  context.drawImage(image, 0, 0, canvas.width, canvas.height);
  const pixels = context.getImageData(0, 0, canvas.width, canvas.height);
  const thumbHashBinary = rgbaToThumbHash(pixels.width, pixels.height, pixels.data);
  const thumbHashBase64 = binaryToBase64(thumbHashBinary);
  const thumbHashDataURL = await thumbHashToDataURL(thumbHashBinary);
  return {
    dataBase64: thumbHashBase64,
    dataUrl: thumbHashDataURL,
    width: resizedWidth,
    height: resizedHeight,
    originalWidth: width,
    originalHeight: height
  };
}
function createThumbHashDataFromThumbHash(rootDir, assetDir, baseUrl, imageFileName, imageFullFileName, imageFullHash, imageFileHash, thumbHash) {
  const fileName = relative(rootDir, imageFileName);
  const thumbhashData = {
    ...thumbHash,
    assetFileName: normalizePath(relative(rootDir, imageFullFileName)),
    assetFullFileName: normalizePath(imageFullFileName),
    assetFullHash: imageFullHash,
    assetFileHash: imageFileHash,
    assetUrl: normalizePath(join(assetDir, fileName)),
    assetUrlWithBase: normalizePath(join(baseUrl, assetDir, fileName))
  };
  if (!thumbhashData.assetUrlWithBase.startsWith("/"))
    thumbhashData.assetUrlWithBase = `/${thumbhashData.assetUrlWithBase}`;
  return thumbhashData;
}
function ThumbnailHashImages() {
  return {
    name: "@nolebase/vitepress-plugin-thumbnail-hash/images",
    enforce: "pre",
    config() {
      return {
        optimizeDeps: {
          exclude: [
            "@nolebase/vitepress-plugin-thumbnail-hash/client"
          ]
        },
        ssr: {
          noExternal: [
            "@nolebase/vitepress-plugin-thumbnail-hash"
          ]
        }
      };
    },
    async configResolved(config) {
      const root = config.root;
      const vitepressConfig = config.vitepress;
      const startsAt = Date.now();
      const moduleNamePrefix = cyan("@nolebase/vitepress-plugin-thumbnail-hash/images");
      const grayPrefix = gray(":");
      const spinnerPrefix = `${moduleNamePrefix}${grayPrefix}`;
      const spinner = ora(`${spinnerPrefix} Prepare to generate hashes for images...`).start();
      const cacheDir = join(vitepressConfig.cacheDir, "@nolebase", "vitepress-plugin-thumbnail-hash", "thumbhashes");
      await mkdir(cacheDir, { recursive: true });
      await rm(join(cacheDir, "*"), { force: true, recursive: true });
      spinner.text = `${spinnerPrefix} Searching for images...`;
      const files = await glob(`${root}/**/*.+(jpg|jpeg|png)`, { nodir: true });
      spinner.text = `${spinnerPrefix} Calculating thumbhashes for images...`;
      const thumbhashes = await Promise.all(files.map(async (file) => {
        const readImageRawData = await readFile(file);
        const imageFullHash = await hash(readImageRawData, -1);
        const imageFileHash = normalizeBase64(imageFullHash.substring(0, 10));
        const calculatedThumbhashData = await calculateThumbHashForFile(readImageRawData);
        return createThumbHashDataFromThumbHash(
          root,
          vitepressConfig.assetsDir,
          vitepressConfig.site.base,
          file,
          file,
          imageFullHash,
          imageFileHash,
          calculatedThumbhashData
        );
      }));
      spinner.text = `${spinnerPrefix} Aggregating calculated thumbhash data...`;
      const thumbhashMap = {};
      for (const thumbhash of thumbhashes)
        thumbhashMap[thumbhash.assetFileName] = thumbhash;
      spinner.text = `${spinnerPrefix} Writing thumbhash data to cache...`;
      await writeFile(join(cacheDir, "map.json"), JSON.stringify(thumbhashMap, null, 2));
      const elapsed = Date.now() - startsAt;
      spinner.succeed(`${spinnerPrefix} Done. ${gray(`(${elapsed}ms)`)}`);
    }
  };
}

export { ThumbnailHashImages };
